File: client/typescript/gamium/src/common/gamium-client.ts
import { ActionChain } from "./actions/action-chain";
import { KeyBy } from "./actions/key-by";
import { GamiumError } from "./errors/gamium-error";
import {
  createExecuteRpc,
  createFindObjects,
  createHello,
  createQueryProfile,
  createQueryScreen,
  GamiumService,
} from "./gamium-service";
import { Inspector } from "./inspect/inspector";
import { delay, stringify, stringifyAllProps } from "./internal/functions";
import { Printable } from "./internal/logs";
import { waitGeneric } from "./internal/utils/wait";
import { Locator } from "./locator/locator";
import { RpcLocator } from "./locator/rpc-locator";
import { Player } from "./object/player";
import {
  DefaultExecuteRpcOptions,
  ExecuteRpcOptions,
} from "./options/execute-rpc-options";
import {
  DefaultFindObjectOptions,
  FindObjectOptions,
} from "./options/find-object-options";
import {
  DefaultSendKeyOptions,
  SendKeyOptions,
} from "./options/send-key-options";
import { DefaultWaitOptions, WaitOptions } from "./options/wait-options";
import {
  ErrorCode,
  HelloResultT,
  ObjectInfoT,
  ObjectLocatorT,
} from "./protocols/generated";
import {
  ObjectInfo,
  QueryProfileResult,
  QueryScreenResult,
} from "./protocols/types";
import { tryify, TryResult } from "./try";
import { UI } from "./ui/ui";
import { Version } from "./version";
import { WaitCondition } from "./wait-condition";

export class GamiumClient {
  constructor(
    private readonly gamiumService: GamiumService,
    readonly logger: Printable
  ) {}

  get connected(): boolean {
    return this.gamiumService.connected;
  }

  async connect(tryCount = 30): Promise<void> {
    this.logger.info(`GamiumClient.connect`);
    await this.gamiumService.connect(tryCount);
  }

  disconnect(): void {
    this.logger.info(`GamiumClient.disconnect`);
    this.gamiumService.disconnect();
  }

  async screen(): Promise<QueryScreenResult> {
    this.logger.info(`GamiumClient.screen`);
    return await this.gamiumService.request(createQueryScreen({}));
  }

  async hello(): Promise<HelloResultT> {
    return await this.gamiumService.request(createHello({ version: Version }), {
      timeout: 1000,
    });
  }

  async profile(): Promise<QueryProfileResult> {
    this.logger.info(`GamiumClient.profile`);
    return await this.gamiumService.request(createQueryProfile({}));
  }

  // object

  async find(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<ObjectInfo> {
    const infos = await this.finds(locator, options);
    if (infos.length === 0) {
      throw new GamiumError(
        ErrorCode.ObjectNotFound,
        `GamiumClient.findObject By:${locator.by}, str:${locator.str} not found`
      );
    }
    const info = infos[0];
    if (info === undefined) {
      throw new GamiumError(
        ErrorCode.ObjectNotFound,
        `GamiumClient.findObject By:${locator.by}, str:${locator.str} not found`
      );
    }
    return info;
  }

  async tryFind(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<TryResult<ObjectInfo>> {
    return tryify(this.find(locator, options));
  }

  async finds(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<ObjectInfo[]> {
    this.logger.info(
      `GamiumClient.findObjects By:${locator.by}, str:${locator.str}`
    );
    const optionMixed: FindObjectOptions = {
      ...DefaultFindObjectOptions(),
      ...options,
    };
    if (
      locator.str === undefined ||
      locator.str === null ||
      locator.str.length === 0
    ) {
      throw new GamiumError(
        ErrorCode.InvalidParameter,
        `GamiumClient.findObject By:${locator.by}, str:${locator.str} is invalid`
      );
    }
    const res = await this.gamiumService.request(
      createFindObjects({
        locator: new ObjectLocatorT(locator.by, locator.str),
      })
    );

    await delay(optionMixed.delayMs);
    const infos: ObjectInfo[] = res.infos.map((info: ObjectInfoT) => {
      return {
        path: info.path?.toString() ?? "",
        name: info.name?.toString() ?? "",
        type: info.type,
        tag: info.tag,
        isActive: info.isActive,
        screenPosition: info.screenPosition ?? { x: 0, y: 0, z: 0 },
        screenRectSize: info.screenRectSize ?? { x: 0, y: 0 },
        position: info.position ?? { x: 0, y: 0, z: 0 },
        rotation: info.rotation ?? { x: 0, y: 0, z: 0, w: 0 },
        text: info.text?.toString() ?? "",
      };
    });
    return infos;
  }

  async tryFinds(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<TryResult<ObjectInfo[]>> {
    return tryify(this.finds(locator, options));
  }

  // action

  actions(): ActionChain {
    return new ActionChain(this.gamiumService);
  }

  async sendKey(
    by: KeyBy,
    options: Partial<SendKeyOptions> = DefaultSendKeyOptions()
  ): Promise<void> {
    return await this.sendKeys([by], options);
  }

  async sendKeys(
    byList: KeyBy[],
    options: Partial<SendKeyOptions> = DefaultSendKeyOptions()
  ): Promise<void> {
    this.logger.info(`GamiumClient.sendKeys codes:${stringify(byList)}`);
    await this.actions().sendKeys(byList, options).perform();
  }

  // execute

  async executeRpc(
    locator: RpcLocator,
    option: Partial<ExecuteRpcOptions> = DefaultExecuteRpcOptions()
  ): Promise<undefined | number | string | object> {
    this.logger.info(
      `GamiumClient.executeRpc By:${locator.by}, className:${
        locator.className
      }, targetName:${locator.targetName}, params:${stringifyAllProps(
        locator.params
      )}`
    );
    const params = locator.params.map((p) => {
      const doc = JSON.stringify(p);
      if (option.castNumberToFloat) {
        return doc.replace(
          /(-?\d+)(\.\d+)?/g,
          (match: string, p1: string, p2: string) => {
            if (p2) {
              return p1 + p2;
            }
            return p1 + ".0";
          }
        );
      }
      return doc;
    });
    const res = await this.gamiumService.request(
      createExecuteRpc({
        by: locator.by,
        className: locator.className,
        targetName: locator.targetName,
        paramDocuments: params,
      })
    );
    if (!res.document || res.document === "null") {
      return undefined;
    }
    const parsed = JSON.parse(res.document.toString()) as unknown;
    const parsedNumber = parsed as number;
    if (null !== parsedNumber) {
      return parsedNumber;
    }
    const parsedString = parsed as string;
    if (null !== parsedString) {
      return parsedString;
    }

    const parsedObject = parsed as object;
    if (null !== parsedObject) {
      return parsedObject;
    }
    throw new GamiumError(
      ErrorCode.ExecuteRpcNotSupportedType,
      `executeRpc By:${locator.by}, className:${
        locator.className
      }, targetName:${
        locator.targetName
      } result:${res.document.toString()} not supported type`
    );
  }

  // player
  async player(
    locator: Locator,
    options: Partial<FindObjectOptions> = DefaultFindObjectOptions()
  ): Promise<Player> {
    const info = await this.find(locator, options);
    return new Player(this, this.gamiumService, info);
  }

  // ui

  ui(): UI {
    return new UI(this, this.gamiumService);
  }

  // inspector

  inspector(): Inspector {
    return new Inspector(this, this.gamiumService);
  }

  async sleep(millseconds: number): Promise<void> {
    this.logger.info(`GamiumClient.sleep millseconds:${millseconds}`);
    await this.actions().sleep(millseconds).perform();
  }

  async wait<Type>(
    condition: WaitCondition<Type>,
    option: Partial<WaitOptions> = DefaultWaitOptions()
  ): Promise<Type> {
    return await waitGeneric(this, condition, option);
  }

  async tryWait<Type>(
    condition: WaitCondition<Type>,
    option: Partial<WaitOptions> = DefaultWaitOptions()
  ): Promise<TryResult<Type>> {
    return tryify(this.wait(condition, option));
  }
}

File: client/python/gamium/gamium/gamium_client.py
import functools
import json
import time
from typing import Any, List, Optional
from gamium.condition.wait_condition import WaitCondition
from gamium.options.wait_options import WaitOptions
from gamium.protocol.generated.Types import ErrorCode
from gamium.utils.generics import T
from gamium.utils.try_utils import TryResult, tryify
from gamium.utils.wait import wait_generic
from gamium.protocol.types import ObjectInfo, Vector2, Vector3, Vector4

from gamium.options import FindObjectOptions, ExecuteRpcOptions, SendKeyOptions

from gamium.igamium_client import IGamiumClient
from gamium.locator.rpc_locator import RpcLocator
from gamium.object.player import Player
from gamium.ui.ui import UI
from gamium.actions.key_by import KeyBy
from gamium.actions.action_chain import ActionChain
from gamium.igamium_service import *
from gamium.errors.gamium_error import GamiumError
from gamium.internal.logger import Logger
from gamium.locator.locator import Locator


class GamiumClient(IGamiumClient):
    def __init__(
        self,
        service: IGamiumService,
        printable: Any = None,
    ):
        self.__internal_loger = Logger()
        if printable:
            self._logger.set_handler(printable)
        self._service = service

    def is_connected(self) -> bool:
        return self._service.is_connected()

    def connect(self) -> None:
        self._logger.info("GamiumClient.connect")
        self._service.connect()

    def screen(self) -> QueryScreenResultT:
        self._logger.info("GamiumClient.screen")
        return self._service.request(create_query_screen())

    def hello(self) -> HelloResultT:
        self._logger.info("GamiumClient.hello")
        return self._service.request(create_hello())

    def profile(self) -> QueryProfileResultT:
        self._logger.info("GamiumClient.profile")
        return self._service.request(create_profile())

    def find(
        self,
        locator: Locator,
        options: FindObjectOptions = FindObjectOptions(),
    ) -> ObjectInfo:
        infos = self.finds(locator, options)
        if 0 == len(infos):
            raise GamiumError(
                ErrorCode.ObjectNotFound,
                f"GamiumClient.find By: {locator.by}, str: {locator.str} not found",
            )
        if infos[0] is None:
            raise GamiumError(
                ErrorCode.ObjectNotFound,
                f"GamiumClient.find By: {locator.by}, str: {locator.str} not found",
            )
        return infos[0]

    def finds(
        self,
        locator: Locator,
        options: FindObjectOptions = FindObjectOptions(),
    ) -> List[ObjectInfo]:
        self._logger.info(f"GamiumClient.finds By: {locator.by}, str: {locator.str}")
        if locator.str is None or 0 == len(locator.str):
            raise GamiumError(
                ErrorCode.InvalidParameter,
                f"GamiumClient.finds By: {locator.by}, str: {locator.str} is invalid",
            )

        fbs_locator = ObjectLocatorT()
        fbs_locator.by = locator.by
        fbs_locator.str = locator.str
        res = self._service.request(create_find_objects(fbs_locator))

        time.sleep(options.delay_ms / 1000)
        infos = []
        for info in res.infos:
            infos.append(
                ObjectInfo(
                    info.path,
                    info.name,
                    info.type,
                    info.tag,
                    info.isActive,
                    Vector3(info.screenPosition.x, info.screenPosition.y, info.screenPosition.z) if info.screenPosition is not None else Vector3.zero(),
                    Vector2(info.screenRectSize.x, info.screenRectSize.y) if info.screenRectSize is not None else Vector2.zero(),
                    Vector3(info.position.x, info.position.y, info.position.z) if info.position is not None else Vector3.zero(),
                    Vector4(info.rotation.x, info.rotation.y, info.rotation.z, info.rotation.w) if info.rotation is not None else Vector4.zero(),
                    info.text,
                )
            )

        return infos

    def actions(self) -> ActionChain:
        return ActionChain(self._service)

    def send_key(self, by: KeyBy, options: SendKeyOptions = SendKeyOptions()) -> None:
        self.send_keys([by], options)

    def send_keys(self, by_list: List[KeyBy], options: SendKeyOptions = SendKeyOptions()) -> None:
        self._logger.info(f"GamiumClient.send_keys {[by.str for by in by_list]}")
        self.actions().send_keys(by_list, options).perform()

    def execute_rpc(
        self,
        locator: RpcLocator,
        option: ExecuteRpcOptions = ExecuteRpcOptions(),
    ) -> Any:
        self._logger.info(f"GamiumClient.execute_rpc By: {locator.by}, class: {locator.class_name}, target: {locator.target_name}")
        params: List[str] = []
        for param in locator.params:
            if option.cast_number_to_float:
                if isinstance(param, int):
                    params.append(str(float(param)))
                    continue
            params.append(str(param))
        res = self._service.request(create_execute_rpc(locator.by, locator.class_name, locator.target_name, params))
        if res.document is None:
            return None
        parsed = json.loads(res.document)
        return parsed

    def player(
        self,
        locator: Locator,
        options: FindObjectOptions = FindObjectOptions(),
    ) -> Player:
        info = self.find(locator, options)
        return Player(self, self._service, info)

    def ui(self) -> UI:
        return UI(
            self,
            self._service,
        )

    def inspector(self):
        # TODO
        pass

    def sleep(self, ms: int) -> None:
        self._logger.info(f"GamiumClient.sleep {ms} ms")
        self.actions().sleep(ms).perform()

    def wait(
        self,
        condition: WaitCondition[T],
        options: WaitOptions = WaitOptions(),
    ) -> T:
        return wait_generic(self, condition, options)

    def try_wait(
        self,
        condition: WaitCondition[T],
        options: WaitOptions = WaitOptions(),
    ) -> TryResult[T]:
        return tryify(functools.partial(self.wait, condition, options))  # type: ignore  # Ignore type error on this line

    @property
    def _logger(self) -> Logger:
        return self.__internal_loger
